### PR TITLE
FEM-1191: Offline and switch player fixes

### PR DESCRIPTION
**Pull request by AntonAFA**
LocalAssetsManager refactoring:

+ Some blocks of code moved to separate functions. for example: checkIfParamsAreValid(mediaSource.getUrl(), localAssetPath, assetId);
+ Added methods that will handle the asset registration/unregistration flow for clear sources.
+ Also added separate methods for registration/unregistration for drm and clear sources.
+ Now, LocalDataStore will store also the PKMediaFormat as byte[] value. The key for this will be build based on asset id with prefix = "assetId:" and suffix = "_type:" in order to make it more informative.
+ Instead of guessingAssetScheme now we build it based on the media format that is saved in LocalDataStore

PlayerController:

+ Some blocks of code moved to separate functions.
+ Added support for change media between different types of player engines (MediaWrapper/ExoplayerWrapper). Now in prepare method we will check if player engines should be switched and if so we will first remove existed view from the hierarchy and after that initialize new player engine.

### CODE REVIEW

File: playkit/src/main/java/com/kaltura/playkit/LocalAssetsManager.java [hunk 1]
```diff
@@ -73,98 +78,116 @@ public LocalAssetsManager(Context context) {
...
      */
     public void registerAsset(@NonNull final PKMediaSource mediaSource, @NonNull final String localAssetPath,
-                                 @NonNull final String assetId, final AssetRegistrationListener listener) {
+                              @NonNull final String assetId, final AssetRegistrationListener listener) {
 
-        // Preflight: check that all parameters are valid.
-        checkNotEmpty(mediaSource.getUrl(), "mediaSource.url");
-        checkNotEmpty(localAssetPath, "localAssetPath");
-        checkNotEmpty(assetId, "assetId");
+        checkIfParamsAreValid(mediaSource.getUrl(), localAssetPath, assetId);
```
Comment (noamtamim) [inline playkit/src/main/java/com/kaltura/playkit/LocalAssetsManager.java]: Why?

File: playkit/src/main/java/com/kaltura/playkit/LocalAssetsManager.java [hunk 2]
```diff
@@ -25,10 +25,12 @@
 public class LocalAssetsManager {
 
     private static final PKLog log = PKLog.get("LocalAssetsManager");
+    private static final String ASSET_ID_PREFIX = "assetId:";
```
Comment (noamtamim) [inline playkit/src/main/java/com/kaltura/playkit/LocalAssetsManager.java]: So what's the final name?

File: playkit/src/main/java/com/kaltura/playkit/LocalAssetsManager.java [hunk 3]
```diff
@@ -73,98 +78,116 @@ public LocalAssetsManager(Context context) {
...
             listener.onFailed(localAssetPath, new Exception("Can't register/refresh when offline"));
             return;
         }
 
-        final PKDrmParams drmParams = findSupportedDrmParams(mediaSource);
-        
+        PKDrmParams drmParams = findSupportedDrmParams(mediaSource);
+
+        //Convert the media format to byte[] in order to store it in the LocalDataStore.
+        byte[] mediaFormat = mediaSource.getMediaFormat().toString().getBytes();
```
Comment (noamtamim) [inline playkit/src/main/java/com/kaltura/playkit/LocalAssetsManager.java]: Null protection if getMediaFormat can be null.

File: playkit/src/main/java/com/kaltura/playkit/LocalAssetsManager.java [hunk 4]
```diff
@@ -73,98 +78,116 @@ public LocalAssetsManager(Context context) {
...
-                                @Override
-                                public void run() {
-                                    listener.onFailed(localAssetPath, e);
-                                }
-                            });
-                        }
-                    }
-                }
-            });
+            registerDrmAsset(localAssetPath, assetId, mediaFormat, drmParams, listener);
```
Comment (noamtamim) [inline playkit/src/main/java/com/kaltura/playkit/LocalAssetsManager.java]: Maybe rename methods to asyncRegisterXYZ to make it clear.

File: playkit/src/main/java/com/kaltura/playkit/LocalAssetsManager.java [hunk 5]
```diff
@@ -233,9 +256,64 @@ public void run() {
...
+        }
+    }
+
+    private static PKDrmParams findSupportedDrmParams(@NonNull PKMediaSource mediaSource) {
+        if (mediaSource.getDrmData() == null) {
+            return null;
+        }
+
+        for (PKDrmParams params : mediaSource.getDrmData()) {
+            switch (params.getScheme()) {
```
Comment (noamtamim) [inline playkit/src/main/java/com/kaltura/playkit/LocalAssetsManager.java]: Scheme enum now has an isSupported method, use it.

File: playkit/src/main/java/com/kaltura/playkit/LocalAssetsManager.java [hunk 6]
```diff
@@ -233,9 +256,64 @@ public void run() {
...
+            String formatName = new String(localDataStore.load(buildAssetKey(assetId)));
+            format = PKMediaFormat.valueOf(formatName);
+        } catch (FileNotFoundException e) {
+            log.e(e.getMessage());
+            return null;
+        }
+
+        switch (format) {
+            case dash:
+                return PKDrmParams.Scheme.WidevineCENC;
```
Comment (noamtamim) [inline playkit/src/main/java/com/kaltura/playkit/LocalAssetsManager.java]: Dash can also be PlayReady scheme, you must store the scheme too. I suggest that the storage entry contains "format:scheme", for example "dash:WidevineCENC".

File: playkit/src/main/java/com/kaltura/playkit/LocalAssetsManager.java (hunk 2 above)
Comment (AntonAFA) [inline playkit/src/main/java/com/kaltura/playkit/LocalAssetsManager.java]: final name will be "assetId:123XYZ_type:" and the value of the key will be the type.

File: playkit/src/main/java/com/kaltura/playkit/LocalAssetsManager.java (hunk 1 above)
Comment (AntonAFA) [inline playkit/src/main/java/com/kaltura/playkit/LocalAssetsManager.java]: why not. separate method for validation block.

Comment (AntonAFA) [inline playkit/src/main/java/com/kaltura/playkit/LocalAssetsManager.java]: logically it is the same.

**Pull request merged**